t/t1092-sparse-checkout-compatibility.sh

```diff
@@ -741,7 +741,14 @@ test_expect_success 'checkout-index with folders' '
 	test_all_match test_must_fail git checkout-index -f -- deep/ &&
 
 	# Outside checkout definition
-	test_all_match test_must_fail git checkout-index -f -- folder1/
+	# Note: although all tests fail (as expected), the messaging differs. For
+	# non-sparse index checkouts, the error is that the "file" does not appear
+	# in the index; for sparse checkouts, the error is explicitly that the
+	# entry is a sparse directory.
```

Comment:
Is it possible to make both implementations say "file is outside the sparse-checkout definition"?

* The "file does not appear in the index" is not entirely correct for full indexes.
* The "file is a sparse directory" has a bit too much internal information.

I wonder if using `if (!path_in_sparse_checkout(name, &the_index))` somewhere would help here.

It's perhaps a bit tricky how this loop is structured, since it is doing a prefix match on a range of index entries.

In this test script, what happens with this example?

```
git sparse-checkout set deep/deeper1
git checkout-index -f -- deep/
```

Note that `deep/deeper2/` is a sparse directory within the `deep/` directory, but we would still want the other entries to succeed, silently ignoring the `deep/deeper2/` entry. This would only be an _error_ if the user ran `git checkout-index -f -- deep/deeper2/`, right?

Reply:
I think it's just comparing the names, not performing a prefix match (the first condition in this check makes sure the names are the same length):

```
if (ce_namelen(ce) != namelen ||
		    memcmp(ce->name, name, namelen))
```

The result is that, from your example, `git checkout-index -f -- deep/` (or any path that doesn't refer to a file entry in the index) returns `git checkout-index: <name> is not in the cache`. It seems like this command expects _exact_ names of cache entries (except for the subcommands using a pathspec, but they handle things differently anyway), in which case `<name> is not in the cache` makes some amount of sense? Agreed that it's misleading, though - would that be something that should be clarified in the command doc, or in this message itself?

As for checking whether something is in the sparse checkout, this doesn't actually block checking out files outside the sparse definition (you can successfully run `git checkout-index -- folder1/a` without any other repo setup). The issue is that sparse directories _do_ appear as standalone cache entries, so without the `S_ISSPARSEDIR` check the entry would be passed along to `checkout_entry` (which then fails because it doesn't handle directories). Since there's no support `checkout-index`-ing directories otherwise, this probably needs to be blocked rather than worked around. Maybe the message should be the same as whenever anyone tries to `checkout-index` a directory in general? I.e., "lie" and say `<name> is not in the cache`, or some other error message that's the same between them?